modules/capture-video/src/main/native/linux/src/capture.c

```diff
@@ -0,0 +1,218 @@
+
+                buffers[n_buffers].length = buf.length;
+                buffers[n_buffers].start = v4l2_mmap(NULL, buf.length,
+                              PROT_READ | PROT_WRITE, MAP_SHARED,
+                              fd, buf.m.offset);
+
+                if (MAP_FAILED == buffers[n_buffers].start) {
+                        perror("mmap");
+                        exit(EXIT_FAILURE);
+                }
+        }
+        for (i = 0; i < n_buffers; ++i) {
+                CLEAR(buf);
+                buf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+                buf.memory = V4L2_MEMORY_MMAP;
+                buf.index = i;
+                xioctl(fd, VIDIOC_QBUF, &buf);
+        }
+        type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+
+        xioctl(fd, VIDIOC_STREAMON, &type);
+        while (active > 0) {
+               do {
+                        FD_ZERO(&fds);
+                        FD_SET(fd, &fds);
+
+                        /* Timeout. */
+                        tv.tv_sec = 2;
+                        tv.tv_usec = 0;
+
+                        r = select(fd + 1, &fds, NULL, NULL, &tv);
+                } while ((r == -1 && (errno = EINTR)));
+                if (r == -1) {
+                        perror("select");
+                        return errno;
+                }
+
+                CLEAR(buf);
+                buf.type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+                buf.memory = V4L2_MEMORY_MMAP;
+                xioctl(fd, VIDIOC_DQBUF, &buf);
+
+                int width = fmt.fmt.pix.width;
+                int height = fmt.fmt.pix.height;
+                void* rgba = malloc(width*height*4);
+                rgbToRgba(rgba, buffers[buf.index].start, width * height);
+    int len = width * height * 4;
+// rgbToRgba(rgba, buffers[buf.index].start, width * height);
+    // memcpy(snd, buffers[buf.index].start, len);
+sendPicturesResult(width, height, 2, rgba, len);
+free(rgba);
+                xioctl(fd, VIDIOC_QBUF, &buf);
+        }
+
+        type = V4L2_BUF_TYPE_VIDEO_CAPTURE;
+        xioctl(fd, VIDIOC_STREAMOFF, &type);
+        for (i = 0; i < n_buffers; ++i)
+                v4l2_munmap(buffers[i].start, buffers[i].length);
+        v4l2_close(fd);
+return 0;
+
+}
+
+int deviceId = -1;
+
+JNIEXPORT void JNICALL Java_com_gluonhq_attachextendedmac_capturevideo_impl_DesktopCaptureVideoService_initCaptureVideo (JNIEnv *env, jclass jClass) {
+    fprintf(stderr, "initCaptureVideo\n");
+    jCaptureVideoServiceClass = (*env)->NewGlobalRef(env, (*env)->FindClass(env, "com/gluonhq/attachextendedmac/capturevideo/impl/DesktopCaptureVideoService"));
+    jCaptureVideoService_setResult = (*env)->GetStaticMethodID(env, jCaptureVideoServiceClass, "setResult", "(III[B)V");
+initializeGrabber();
+}
+
+JNIEXPORT void JNICALL Java_com_gluonhq_attachextendedmac_capturevideo_impl_DesktopCaptureVideoService_nativeStop (JNIEnv *env, jclass jClass) {
+    fprintf(stderr, "LINUXNATIVESTOP\n");
```

Comment:
To stop the camera, `active = 0`?
(making sure that frames are not sent to the java layer if active != 1)

Reply:
there need to be more resource working and ioctl to make sure the devices are in the idle state again